# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

[Volver a Contenidos](#back)

# Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importar pandas
import pandas as pd 

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv('/datasets/music_project_en.csv') 

Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtener las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [4]:
# Obtener la información general sobre nuestros datos
print(df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`.
    'userID' no sigue el formato snake_case.




### Conclusiones

El dataset incluye 7 columnas y 67059 filas, cuyos datos corresponden a tipo object (o string). Para determinar si la cantidad de datos sirve para corroborar la hipótesis de manera fiable, habría que conocer la cantidad de habitantes promedio por ciudad. Los datos almacenados en cada columna se pueden entender mediante la documentación proporcionada. 
Al aplicar el método .info() nos damos cuenta que las columnas 'track', 'artist' y 'genre' poseen valores nulos. Estos valores afectarían los resultados si quisiéramos analizar los las canciones o artistas más populares o saber la tendencia de género por día, por ejemplo. 

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
# Bucle en los encabezados poniendo todo en minúsculas
lowered_col_names = []

for old_name in df.columns:
    name_lowered = old_name.lower()
    lowered_col_names.append(name_lowered)

df.columns = lowered_col_names

Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [7]:
# Bucle en los encabezados eliminando los espacios
no_spaces_col_names = []

for lowered_name in df.columns:
    name_no_spaces = lowered_name.strip()
    no_spaces_col_names.append(name_no_spaces)

df.columns = no_spaces_col_names

Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
# Cambiar el nombre de la columna "userid"
column_new = {
    'userid': 'user_id'        
}

df.rename(columns=column_new, inplace=True)

Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
# Comprobar el resultado: la lista de encabezados
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [10]:
# Calcular el número de valores ausentes
print(df.isna().sum())


user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [11]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace = ['track', 'artist', 'genre']

for col in columns_to_replace:
    df[col].fillna('unknown', inplace=True)

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [12]:
# Contar valores ausentes
(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [13]:
# Contar duplicados explícitos
df.duplicated().sum()

3826

Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
# Eliminar duplicados explícitos
df = df.drop_duplicates().reset_index(drop=True)

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
# Comprobar de nuevo si hay duplicados
df.duplicated().sum()


0

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
# Inspeccionar los nombres de géneros únicos
df['genre'].sort_values().unique()

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'ambient', 'americana', 'animated', 'anime',
       'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook',
       'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom',
       'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian',
       'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop',
       'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber',
       'children', 'chill', 'chinese', 'choral', 'christian', 'christmas',
       'classical', 'classicmetal', 'club', 'colombian', 'comedy',
       'conjazz', 'contemporary', 'country', 'cuban', 'dance',
       'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock',
       'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat',
       'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy',
       'electronic', 'electropop', 'emo', 'entehno', 'epicmetal',
       'estrada', 'ethnic', 'eurofo

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [17]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(df, column, wrong_values, correct_value):
    for wrong_value in wrong_values:
        df[column] = df[column].replace(wrong_value, correct_value)
    return df        

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [18]:
# Eliminar duplicados implícitos
wrong_genres = ['hip', 'hop', 'hip-hop']
correct_genre = ('hiphop')

replace_wrong_genres(df, 'genre', wrong_genres, correct_genre)

user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
61248  729CBB09                            My Name            McLean   
61249  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
61250  C5E3A0D5                          Jalopiina           unknown   
61251  321D0506                      Freight Train     Chas McDevitt   
61252  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
61248         rnb  Springfield  13:32:28  Wednesday  
61249      hiphop  Shelbyville  10:00:00     Monday  
61250  industrial  Springfield  20:09:26     Friday  
61251        rock  Springfield  21:43:59     Friday  
61252     country  Springfield  21:59:46     Friday  

[61253 rows x 7 columns]

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# Comprobación de duplicados implícitos
df['genre'].sort_values().unique()

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'ambient', 'americana', 'animated', 'anime',
       'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook',
       'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom',
       'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian',
       'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop',
       'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber',
       'children', 'chill', 'chinese', 'choral', 'christian', 'christmas',
       'classical', 'classicmetal', 'club', 'colombian', 'comedy',
       'conjazz', 'contemporary', 'country', 'cuban', 'dance',
       'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock',
       'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat',
       'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy',
       'electronic', 'electropop', 'emo', 'entehno', 'epicmetal',
       'estrada', 'ethnic', 'eurofo

[Volver a Contenidos](#back)

**Observaciones pre-procesamiento de datos**

Se utilizó método .drop_duplicates() para obtener el siguiente resultado:
  1. Los datos duplicados son de 3826, así que la cantidad total de datos a analizar es de 61.252. Por lo que la muestra es muy pequeña para obtener conclusiones fiables
    
Para contabilizar los valores ausentes, se utilizó el métod .isna() en conjunto con .sum(). Los resultados son los siguientes:

  1. Dentro de los datos, la columna 'artist' presente 7567 valores ausente. A modo de conclusión, si se quisiera analizar el artista más popular dentro de la población, este resultado estaría alterado. 
  2. La columna 'genre' presenta 1196 valores ausente, por lo que afecta directamente la conclusión de la hipótesis. Esos valores cambiarán la tendencia en preferencias musicales.
   

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [20]:
# Contar las canciones reproducidas en cada ciudad
df.groupby('city')['track'].count()

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

Al agrupar los datos por ciudad, nos damos cuenta que Shelbyville escuchó 18512 y Springfield 42741, lo que representan un 30.2% y 69.8% del total de los datos respectivamente. 

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [21]:
# Calcular las canciones reproducidas en cada uno de los tres días
df.groupby('day')['track'].count()

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

Observando los datos, se puede deducir que el inicio y término de la semana laboral son los días en que los habitantes de a,bas ciudades prefieren escuchar música, con tan solo una leve mayoría de canciones reproducidas el día Viernes. 
Miércoles es un día donde se escucha menos música en las ciudades.


Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [22]:
# Declara la función number_tracks() con dos parámetros: day= y city=.

    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=

    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=

    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()

    # Devolve el número de valores de la columna 'user_id'
    
def number_tracks(day, city):
    data = df[df['day'] == day]
    data = data[data['city'] == city] 
    count = data['user_id'].count()
    return count  
    

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [23]:
# El número de canciones reproducidas en Springfield el lunes
number_tracks('Monday', 'Springfield')

15740

In [24]:
# El número de canciones reproducidas en Shelbyville el lunes
number_tracks('Monday', 'Shelbyville')

5614

In [25]:
# El número de canciones reproducidas en Springfield el miércoles
number_tracks('Wednesday', 'Springfield')

11056

In [26]:
# El número de canciones reproducidas en Shelbyville el miércoles
number_tracks('Wednesday', 'Shelbyville')

7003

In [27]:
# El número de canciones reproducidas en Springfield el viernes
number_tracks('Friday', 'Springfield')

15945

In [28]:
# El número de canciones reproducidas en Shelbyville el viernes
number_tracks('Friday', 'Shelbyville')

5895

**Conclusiones**

La hipótesis es correcta. 
En primer lugar, la cantidad de canciones que se escucha en Springfield es significativamente mayor a la escuchada en Shelbyville. 
Por otro lado, los días con mayor reproducción de canciones en Springfield son lunes y viernes, mientras que en Shelbyville esos días se escucha menos música. El día miércoles es el día donde la cantidad de canciones  reproducidas es menor en Springfield, pero el Shelbyville es el día preferido para escuchar música.


[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

El objetivo de este trabajo fue probar la hipótesis acerca de una posible diferencia de usuarios dependiendo de la ciudad y el día. Para esto, se comienza con una descripción de los datos utilizando los métodos head() e info(). Como resultado de esta primera etapa, nos damos cuenta de que las columnas 'track', 'artist' y 'genre' poseen valores nulos.
La segunda fase corresponde al preprocesamiento de datos. En este momento, se estandarizan las columnas, se reemplazan los valores ausentes por ‘unknown’ y se eliminan duplicados explícitos. Además se reemplazan las  variables de ‘hiphop’ y en la columna ‘genre’.

Finalmente, se prueba la hipótesis al comparar las canciones reproducidas por Springfield y Shelbyville para lunes, miércoles y viernes. Como resultado, confirmamos la hipótesis y obtenemos los siguientes insights:   

*Springfield*:
    lunes: 15740,
    miércoles: 11056,
    viernes: 15945.

Día con mayor reproducción de de canciones: viernes.
Día con menor reproducción de de canciones: miércoles.


*Shelbyville*:
    lunes: 5614,
    miércoles: 7003,
    viernes: 5895.

Día con mayor reproducción de de canciones: miércoles.
Día con menor reproducción de de canciones: lunes.

Con los datos recopilados, se puede concluir que los comportamientos de ambas ciudades con respecto a la cantidad de canciones reproducidas durante la semana son muy distintos: Springfield escucha casi dos veces más música que Shelbyville y los comportamientos diarios de cada ciudad son opuestos. 

La falta de datos en 'genre' dificulta saber el comportamiento de las ciudades con respecto a gustos musicales. 

[Volver a Contenidos](#back)